### DISCUSSION TITLE
yarn 2 lockfile update fail

### DISCUSSION THREAD

**Discussion by olegshilov** [renovatebot/renovate#9481] [Request Help] [answered]
Hello everybody.
I want to configure renovate to update the dependencies in my project that uses Yarn 2 PnP.
I faced the following problem, renovate does not want to update the yarn.lock during the dependency update process.
It failed with the following error: `The lockfile would have been modified by this install, which is explicitly forbidden.`

<details>
  <summary>More logs</summary>

```sh
DEBUG: Executing command (repository=olegshilov/modern-frontend-example, branch=renovate/astroturf-1.0.x)
       "command": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\":\"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\""
DEBUG: lock file error (repository=olegshilov/modern-frontend-example, branch=renovate/astroturf-1.0.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\":\"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"",
         "stdout": "\n> yarn@1.22.10 preinstall /home/ubuntu/.npm-global/lib/node_modules/yarn\n> :; (node ./preinstall.js > /dev/null 2>&1 || true)\n\n/home/ubuntu/.npm-global/bin/yarn -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n/home/ubuntu/.npm-global/bin/yarnpkg -> /home/ubuntu/.npm-global/lib/node_modules/yarn/bin/yarn.js\n+ yarn@1.22.10\nadded 1 package in 0.417s\n➤ YN0000: ┌ Resolution step\n➤ YN0000: └ Completed in 0s 352ms\n\n➤ YN0000: ┌ Post-resolution validation\n➤ YN0000: │ @@ -2197,11 +2197,11 @@\n➤ YN0000: │    checksum: 876231688c66400473ba505731df37ea436e574dd524520294cc3bbc54ea40334865e01fa0d074d74d036ee874ee7e62f486ea38bc421ee8e6a871c06f011766\n➤ YN0000: │    languageName: node\n➤ YN0000: │    linkType: hard\n➤ YN0000: │  \n➤ YN0028: │ -\"astroturf@npm:1.0.0-beta.20\":\n➤ YN0028: │ -  version: 1.0.0-beta.20\n➤ YN0028: │ -  resolution: \"astroturf@npm:1.0.0-beta.20\"\n➤ YN0028: │ +\"astroturf@npm:1.0.0-beta.21\":\n➤ YN0028: │ +  version: 1.0.0-beta.21\n➤ YN0028: │ +  resolution: \"astroturf@npm:1.0.0-beta.21\"\n➤ YN0000: │    dependencies:\n➤ YN0000: │      \"@babel/code-frame\": ^7.12.13\n➤ YN0000: │      \"@babel/core\": ^7.13.10\n➤ YN0000: │      \"@babel/generator\": ^7.13.9\n➤ YN0000: │ @@ -2217,9 +2217,9 @@\n➤ YN0000: │      \"@types/webpack\": ^4.41.26\n➤ YN0000: │      chalk: ^4.1.0\n➤ YN0000: │      common-tags: ^1.8.0\n➤ YN0000: │      cosmiconfig: ^7.0.0\n➤ YN0028: │ -    css-loader: ^5.1.2\n➤ YN0028: │ +    css-loader: ^5.1.3\n➤ YN0000: │      fast-levenshtein: ^3.0.0\n➤ YN0000: │      find-cache-dir: ^3.3.1\n➤ YN0000: │      fs-extra: ^9.1.0\n➤ YN0000: │      json5: ^2.2.0\n➤ YN0000: │ @@ -2231,9 +2231,8 @@\n➤ YN0000: │      resolve: ^1.20.0\n➤ YN0000: │      webpack-virtual-modules: ^0.3.2\n➤ YN0000: │    peerDependencies:\n➤ YN0000: │      webpack: \">=2\"\n➤ YN0028: │ -  checksum: 7b6c9240adc28d272f31f14b71f1e70a936270b5c04c4f4dabe6236243fe7a5317a6240dc9e3deaf8d8f906a51a9915d39e6537cb49a181c3db900dba4d5f4af\n➤ YN0000: │    languageName: node\n➤ YN0000: │    linkType: hard\n➤ YN0000: │  \n➤ YN0000: │  \"async@npm:^2.6.2\":\n➤ YN0000: │ @@ -3300,9 +3299,9 @@\n➤ YN0000: │    checksum: 88d891ba18f821e8a94d821ecdd723c606019462664c7d86e7d8731622bd26f9d55582e494bcc2a62f9399cc7b89049ddc8a9d1e8f1bf1a133c2427739d2d334\n➤ YN0000: │    languageName: node\n➤ YN0000: │    linkType: hard\n➤ YN0000: │  \n➤ YN0028: │ -\"css-loader@npm:5.1.3, css-loader@npm:^5.1.2\":\n➤ YN0028: │ +\"css-loader@npm:5.1.3\":\n➤ YN0000: │    version: 5.1.3\n➤ YN0000: │    resolution: \"css-loader@npm:5.1.3\"\n➤ YN0000: │    dependencies:\n➤ YN0000: │      camelcase: ^6.2.0\n➤ YN0000: │ @@ -3322,8 +3321,29 @@\n➤ YN0000: │    checksum: ab9fa688da23d6b43f444522ca41fa446053a166784dddc3f7ca446061af57622a61b1f178d54e63662732e1976a26362d7ccde5b14d1317775f13016572c417\n➤ YN0000: │    languageName: node\n➤ YN0000: │    linkType: hard\n➤ YN0000: │  \n➤ YN0028: │ +\"css-loader@npm:^5.1.3\":\n➤ YN0028: │ +  version: 5.2.1\n➤ YN0028: │ +  resolution: \"css-loader@npm:5.2.1\"\n➤ YN0028: │ +  dependencies:\n➤ YN0028: │ +    camelcase: ^6.2.0\n➤ YN0028: │ +    cssesc: ^3.0.0\n➤ YN0028: │ +    icss-utils: ^5.1.0\n➤ YN0028: │ +    loader-utils: ^2.0.0\n➤ YN0028: │ +    postcss: ^8.2.8\n➤ YN0028: │ +    postcss-modules-extract-imports: ^3.0.0\n➤ YN0028: │ +    postcss-modules-local-by-default: ^4.0.0\n➤ YN0028: │ +    postcss-modules-scope: ^3.0.0\n➤ YN0028: │ +    postcss-modules-values: ^4.0.0\n➤ YN0028: │ +    postcss-value-parser: ^4.1.0\n➤ YN0028: │ +    schema-utils: ^3.0.0\n➤ YN0028: │ +    semver: ^7.3.4\n➤ YN0028: │ +  peerDependencies:\n➤ YN0028: │ +    webpack: ^4.27.0 || ^5.0.0\n➤ YN0028: │ +  languageName: node\n➤ YN0028: │ +  linkType: hard\n➤ YN0028: │ +\n➤ YN0000: │  \"css-prefers-color-scheme@npm:^3.1.1\":\n➤ YN0000: │    version: 3.1.1\n➤ YN0000: │    resolution: \"css-prefers-color-scheme@npm:3.1.1\"\n➤ YN0000: │    dependencies:\n➤ YN0000: │ @@ -11708,9 +11728,9 @@\n➤ YN0000: │      \"@types/react\": 17.0.3\n➤ YN0000: │      \"@types/react-dom\": 17.0.2\n➤ YN0000: │      \"@types/react-redux\": 7.1.16\n➤ YN0000: │      \"@types/redux-logger\": 3.0.8\n➤ YN0028: │ -    astroturf: 1.0.0-beta.20\n➤ YN0028: │ +    astroturf: 1.0.0-beta.21\n➤ YN0000: │      css-loader: 5.1.3\n➤ YN0000: │      cssnano: 4.1.10\n➤ YN0000: │      file-loader: 6.2.0\n➤ YN0000: │      fork-ts-checker-webpack-plugin: 6.1.1\n➤ YN0000: │ @@ -11744,9 +11764,9 @@\n➤ YN0000: │      \"@types/jest\": 26.0.21\n➤ YN0000: │      \"@types/react\": 17.0.3\n➤ YN0000: │      \"@types/react-dom\": 17.0.2\n➤ YN0000: │      \"@types/react-redux\": 7.1.16\n➤ YN0028: │ -    astroturf: 1.0.0-beta.20\n➤ YN0028: │ +    astroturf: 1.0.0-beta.21\n➤ YN0000: │      react: 17.0.1\n➤ YN0000: │      react-dom: 17.0.1\n➤ YN0000: │      react-redux: 7.2.2\n➤ YN0000: │      redux: 4.0.5\n➤ YN0000: │ \n➤ YN0028: │ The lockfile would have been modified by this install, which is explicitly forbidden.\n➤ YN0000: └ Completed\n➤ YN0000: Failed with errors in 0s 454ms\n",
         "stderr": "",
         "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\":\"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"\n",
         "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\":\"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_HTTP_TIMEOUT -e YARN_ENABLE_SCRIPTS -w \"/tmp/renovate/repos/github/olegshilov/modern-frontend-example\" docker.io/renovate/node bash -l -c \"npm i -g yarn && yarn install\"\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:315:20)\n    at ChildProcess.EventEmitter.emit (domain.js:467:12)\n    at maybeClose (internal/child_process.js:1048:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       },
       "type": "yarn"
DEBUG: No updated lock files in branch (repository=olegshilov/modern-frontend-example, branch=renovate/astroturf-1.0.x)
```
</details>

Related workflow with logs: https://github.com/olegshilov/modern-frontend-example/runs/2309749465?check_suite_focus=true

Used configs may found here
- https://github.com/olegshilov/modern-frontend-example/blob/main/config/renovate-self-hosted.json
- https://github.com/olegshilov/modern-frontend-example/blob/main/renovate.json

**Comment by olegshilov** [comment]
All PR's created by renovate has no `yarn.lock` file updates.
https://github.com/olegshilov/modern-frontend-example/pull/7/commits/228eaf71ae3572e141d6c99075f0f67e018a8144

**Reply by alainib** [reply]
hello. where to put `YARN_ENABLE_IMMUTABLE_INSTALLS = false` please ? i got this error today but i create new version each which. i don't get why i get this error now only. 
thanks

**Reply by lolmaus** [reply]
@alainib This is an *environment variable*.

You put it into the beginning of a command that invokes `yarn`:

```
YARN_ENABLE_IMMUTABLE_INSTALLS=false yarn install
```

Or you can put it in a `.env` file if your project uses https://github.com/motdotla/dotenv .

Or you can put it into your CI: either GUI or config file (e. g. `ci.yml`).

***

But instead of using `YARN_ENABLE_IMMUTABLE_INSTALLS=true`, consider resolving the initial problem that causes the necessity to use it.

Make a fresh local clone of your repo and run in your repo. For Yarn 1, use `yarn install --frozen-lockfile --frozen-dependencies`. For Yarn Berry (aka Yarn 2 aka Yarn 3), use `CI=true yarn install`.

See if it passes. If it does, then there is a discrepancy between your original repo and the remote. Try to figure it out and resolve it. For me, the problem was that one of the local subfolders was checked into the repo as a git submodule, so that none of it's files (including a `package.json`) appeared in the remote repo, but they were present locally.

If it also fails, they you want to generate a new lockfile in the freshly cloned repo, commit it and push to the remote. Then pull it into the old local repo and see if the old local repo causes an error with `--frozen-lockfile` or `CI=true`.

**Comment by rarkins** [comment]
It's throwing this error: https://github.com/yarnpkg/berry/blob/95dd7d6fe3e2ddf460a79c529231151d27012044/packages/yarnpkg-core/sources/Project.ts#L1401

It seems to be because "frozen lockfile" is true. Link: https://yarnpkg.com/advanced/error-codes#yn0028---frozen_lockfile_exception

I wonder if this is because we set CI=true: https://github.com/renovatebot/renovate/blob/2a05964f64162218491d869dc4d604969e3b4f3c/lib/manager/npm/post-update/yarn.ts#L82

Looks like Yarn changed this recently: https://github.com/yarnpkg/berry/commit/f958c1fbf91b01c5a6de617508338005f13e3113

FYI @ylemkimon

**Reply by rarkins** [reply]
Strange thing is that it's listed as a breaking change for Yarn 3 in the changelog (i.e. shouldn't apply to Yarn 2): https://github.com/yarnpkg/berry/blob/e64b285fe9e313a5d9f67a502bd6d8a8894ec9df/CHANGELOG.md#L14

@olegshilov is your project using Yarn 2 or 3?

**Reply by olegshilov** [reply]
@rarkins we use yarn 2
```sh
❯ yarn --version
2.4.0-git.20210409.hash-2988e810
```

**Reply by merceyz** [reply]
You can disable the frozen lockfile by setting `YARN_ENABLE_IMMUTABLE_INSTALLS=false`

**Reply by olegshilov** [reply]
@merceyz thanks. this workaround helps me to use the latest yarn version.

**Reply by lolmaus** [reply]
> You can disable the frozen lockfile by setting `YARN_ENABLE_IMMUTABLE_INSTALLS=false`

Thanks! This resolved my problem.

But why is this necessary? Is there a GitHub issue tracking this?

**Reply by rarkins** [reply]
> > You can disable the frozen lockfile by setting `YARN_ENABLE_IMMUTABLE_INSTALLS=false`
> 
> 
> 
> Thanks! This resolved my problem.
> 
> 
> 
> But why is this necessary? Is there a GitHub issue tracking this?
> 
> 

If you can publicly reproduce the problem with the latest Renovate then please create a bug report and we'll debug it.

**Reply by lolmaus** [reply]
> If you can publicly reproduce the problem with the latest Renovate then please create a bug report and we'll debug it.

My issue is with Yarn, not Renovate.

Asking here because it is the *only* page in the whole internet describing a workaround to the problem.

**Reply by rarkins** [reply]
I think it defaults to false on CI so you can be alerted if your lock file is incompatible/invalid with your package.json

**Reply by merceyz** [reply]
> But why is this necessary? Is there a GitHub issue tracking this?

It's on purpose, your CI should have the same dependencies as you had locally otherwise it can hide issues. It was enabled by default in https://github.com/yarnpkg/berry/pull/2530

> Asking here because it is the only page in the whole internet describing a workaround to the problem.

That's funny as it's mentioned in the changelog

> The enableImmutableInstalls will now default to true on CI (we still recommend to explicitly use --immutable on the CLI).
> You can re-allow mutations by adding YARN_ENABLE_IMMUTABLE_INSTALLS=false in your environment variables.
> https://github.com/yarnpkg/berry/blob/e53036a37f5763d2568a1f7465c9a7562e091129/CHANGELOG.md#breaking-changes

**Reply by lolmaus** [reply]
> It's on purpose, your CI should have the same dependencies as you had locally otherwise it can hide issues

When I run `yarn` locally and then `git status`, I can see that `yarn.lock` does not get modified: there are no changes since the commit that was pushed and CI failed with:

> YN0028: The lockfile would have been created by this install, which is explicitly forbidden.

Note how it says "created". I have not been able to google up anything with "created". This error message is only ever mentioned with "modified".

I believe it is a bug in Yarn 3.

**Reply by lolmaus** [reply]
Created an issue: https://github.com/yarnpkg/berry/issues/2948

**Comment by olegshilov** [comment, marked as answer]
Good day gentlemen. I have fresh news here.
The problem was solved by downgrading to the latest stable version of yarn.
Now the lockfiles are generated as expected.
https://github.com/olegshilov/modern-frontend-example/pull/12

**Reply by rarkins** [reply]
Great. FYI I was told that when you take the unstable 2.x it's essentially like taking the new 3.x because it's built off master

**Comment by alainib** [comment]
this is how i solved it after 3 days ...



this is a solution : using `yarn set version 2.4.2`alone don't work

  **.gitlab-ci.yml**

    stages:
      - test
      - release
    
    unit-tests:
      stage: test
      image: node:15.2.1-alpine3.11
      cache:
        key: ${CI_COMMIT_REF_SLUG}
        paths:
          - .yarn/
      script:
        - yarn set version berry      <--- get last version
        - yarn set version 2.4.2      <--- then downgrade to a specific version
        - yarn install
        - yarn build
        - yarn test
        ...

Dockerfile ( for build in CI) 

 

    ...
    ENV NODE_ENV production
    RUN yarn set version berry
    RUN yarn set version 2.4.2
    RUN yarn install
    ...

**Reply by merceyz** [reply]
You should be committing those files and not reinstall it on every run
https://yarnpkg.com/getting-started/install#per-project-install

**Reply by alainib** [reply]
sorry but i don't get what you mean. 
Some files specific to yarn are in `.gitignore` to don't be commit : 
/.pnp.js
/.yarnrc.yml
/.yarn/ 

which file should i commit ? 
thank you

**Reply by merceyz** [reply]
Read the docs I linked to

**Comment by ranjeet-bhati** [comment]
I was getting the same error in gitlab ci and i solved by the below solution:

   stage: build
   image: node:18.12.1
   script: 
     - yarn set version 1.22.19
     - yarn install
